Patch Block.isVisuallyOpaque() to be state-aware

bs2609 opened the pull request:
This adds an override of `isVisuallyOpaque()` that takes an `IBlockState` parameter. Tried to keep changes as small as possible.

bs2609 committed:
Merge remote-tracking branch 'refs/remotes/MinecraftForge/1.10.x' into is-visually-opaque-patches

bs2609 committed:
Patch isVisuallyOpaque() to be state-aware

williewillus commented:
I'd name it properly too, isVisuallyOpaque is a terrible name:

https://github.com/ModCoderPack/MCPBot-Issues/issues/289

bs2609 committed:
Change name of method as per suggestion

bs2609 commented:
OK, renamed method to `causesSuffocation`, and added notes about usage from the [MCP issue](https://github.com/ModCoderPack/MCPBot-Issues/issues/289).

bs2609 commented on patches/minecraft/net/minecraft/block/Block.java.patch:
```diff
@@ -274,12 +274,16 @@
-+     * @return True if the block is visually opaque
++     * @return True if the block can suffocate entities that are inside it.
 +     */
-+    public boolean isVisuallyOpaque(IBlockState state)
```
This is the newly-added version that takes an `IBlockState` parameter. The original still exists and has been marked deprecated, see patch line 33.

mezz commented on patches/minecraft/net/minecraft/block/Block.java.patch:
```diff
@@ -266,6 +274,21 @@
++     * @param state The current state
++     * @return True if the block can suffocate entities that are inside it.
++     */
++    public boolean causesSuffocation(IBlockState state)
```
I'm pretty sure this new function is exactly the same as `Block.func_149637_q`.

bs2609 replied:
That's certainly possible, I need to have a proper look at 1.11 code.
I'll close this once I've done that if Mojang sorted this themselves.

bs2609 replied:
`func_149637_q` is `isBlockNormalCube`. The default logic used is the same, but they are called from different places and overriden differently.

bs2609 replied:
OK, the actual 1.11 methods are `func_191058_s` (`IBlockProperties.isViewBlocking`) and `func_176214_u` (`Block.causesSuffocation`). Mojang didn't update the default implementation though >_>

mezz reviewed (changes requested):
Looks good but I think your new method already exists.

bs2609 commented:
Related: https://github.com/ModCoderPack/MCPBot-Issues/issues/311

mezz commented:
I'm very against adding more "is an opaque block with various micro-differences" methods to `Block.java`, it's already very difficult to navigate.
Is the existing method good enough for this purpose?

bs2609 commented:
What's in 1.11 is Good Enough(tm), the MCP naming is the bigger issue there.

I'll close this, it's a bit too late for 1.10.x now.

Thanks for the review.

mezz commented:
Cool, thanks.
